compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -1594,6 +1594,72 @@ TR::Register *OMR::Power::TreeEvaluator::vstoreEvaluator(TR::Node *node, TR::Cod
    return NULL;
    }
 
+TR::Register *OMR::Power::TreeEvaluator::awrtbarEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   // The wrtbar IL op represents a store with side effects.
+   // Currently we don't use the side effect node. So just evaluate it and decrement the reference count.
+   TR::Node *sideEffectNode = node->getSecondChild();
+   cg->evaluate(sideEffectNode);
+   cg->decReferenceCount(sideEffectNode);
+   // Delegate the evaluation of the remaining children and the store operation to the storeEvaluator.
+   return TR::TreeEvaluator::astoreEvaluator(node, cg);
+   }
+
+TR::Register *OMR::Power::TreeEvaluator::awrtbariEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   // The wrtbar IL op represents a store with side effects.
+   // Currently we don't use the side effect node. So just evaluate it and decrement the reference count.
+   TR::Node *sideEffectNode = node->getThirdChild();
+   cg->evaluate(sideEffectNode);
+   cg->decReferenceCount(sideEffectNode);
+   // Delegate the evaluation of the remaining children and the store operation to the storeEvaluator.
+   return TR::TreeEvaluator::astoreEvaluator(node, cg);  
+   }
+
+TR::Register *OMR::Power::TreeEvaluator::dwrtbarEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   // The wrtbar IL op represents a store with side effects.
+   // Currently we don't use the side effect node. So just evaluate it and decrement the reference count.
+   TR::Node *sideEffectNode = node->getSecondChild();
+   cg->evaluate(sideEffectNode);
+   cg->decReferenceCount(sideEffectNode);
+   // Delegate the evaluation of the remaining children and the store operation to the storeEvaluator.
+   return TR::TreeEvaluator::dstoreEvaluator(node, cg);
```

Comment:
There is some specialized code in `dstoreEvaluator` that handles double stores of `lbits2d` opcodes.  If such a scenario occurs, it looks like it tries to change the identity of the node to an `lstore`, have it evaluated by the `lstoreEvaluator`, and then changes the identity back to a `dstore`.  However, the input node is actually the two-child write barrier node, and the recreate logic will always create a new Node (based on my understanding of the recreate logic).  This is strictly incorrect as a `dstore` should only have one child, yet the number of children will be set to two.

I think there already is an existing problem here, but this code is making it worse because you're now creating two dummy nodes instead of one and you're creating them with more children than is allowed on a `dstore` node.

I am skeptical that this will cause any functional issues, but it is a leak and the node you're creating aren't strictly valid.

Correct me if I'm wrong on any of this...

Reply:
Yeah, those extra children still hanging on the nodes and not used worried me too. @cathyzhyi @dchopra001 @AlenBadel do you think that before calling regular store/load Evaluators we should set the number of children appropriately and perhaps even reset the opcode? 

On another hand, who would produce such trees and not support them?

Reply:
I don't see this `lbits2d` optimization in the Z evaluators for `fstore` and `dstore`.

Reply:
Please raise an issue to track the resolution of this problem.  It is a pre-existing issue but not one that should be left unresolved.

Reply:
Opened an issue : https://github.com/eclipse/omr/issues/4080